<a href="https://colab.research.google.com/github/elchupacabris/DNAShot/blob/main/Firebase_Hole_Data_Upload_Fix.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# @title Default title text
import pandas as pd
from datetime import datetime
import numpy as np

# --- Firebase Admin SDK Initialization ---
# This part is crucial for interacting with Firestore.
# Ensure you have installed the Firebase Admin SDK: pip install firebase-admin
import firebase_admin
from firebase_admin import credentials, firestore

# Initialize Firebase Admin SDK if it hasn't been initialized yet
# Replace 'path/to/your/serviceAccountKey.json' with the actual path to your Firebase service account key
try:
    # Check if a default app is already initialized to avoid re-initialization errors
    firebase_admin.get_app()
except ValueError:
    # If not initialized, create credentials and initialize the app
    # NOTE: Make sure 'serviceAccountKey.json' is in your Colab environment or specify the correct path
    # Ensure your Google Drive is mounted before running this.
    service_account_key_path = '/content/drive/MyDrive/serviceAccountKey.json'
    try:
        cred = credentials.Certificate(service_account_key_path)
        firebase_admin.initialize_app(cred)
        print("✅ Firebase Admin SDK initialized successfully.")
    except FileNotFoundError:
        print(f"❌ Error: Service account key not found at '{service_account_key_path}'. Please check the path and ensure your Google Drive is mounted.")
    except Exception as e:
        print(f"❌ Error initializing Firebase Admin SDK: {e}")


# Get the Firestore client
try:
    db = firestore.client()
    print("✅ Firestore client obtained successfully.")
except NameError:
    print("❌ Error: Firebase Admin SDK not initialized. Cannot get Firestore client.")
except Exception as e:
    print(f"❌ Error getting Firestore client: {e}")


# --- DataFrame Loading ---
# Load your data into deposit_df and index_df
# Use the provided path from Google Drive
file_path = "/content/drive/MyDrive/Raw data/MASTER-FIREBASE.xlsx" # Updated file path

# IMPORTANT: If your data is on different sheets within the same Excel file,
# specify the sheet_name parameter (e.g., sheet_name='HoleDataSheet').
# If they are separate files, update the file paths accordingly.
try:
    # Assuming 'deposit_df' contains the hole-by-hole data
    deposit_df = pd.read_excel(file_path, sheet_name='ALLDATA') # Using 'ALLDATA' as specified
    print("✅ deposit_df loaded successfully.")
except FileNotFoundError:
    print(f"❌ Error: '{file_path}' not found for deposit_df. Please check the file path.")
except Exception as e:
    print(f"❌ Error loading deposit_df: {e}")

try:
    # Assuming 'index_df' contains the scorecard summary data
    index_df = pd.read_excel(file_path, sheet_name='INDEX') # Using 'INDEX' as specified
    print("✅ index_df loaded successfully.")
except FileNotFoundError:
    print(f"❌ Error: '{file_path}' not found for index_df. Please check the file path.")
except Exception as e:
    print(f"❌ Error loading index_df: {e}")


# The following code for data processing and Firestore upload is commented out
# and will be addressed in subsequent steps of the plan.

# # --- User ID Mapping ---
# # Map player names to user IDs
# user_id_map = {
#     'ML': '998b30f58115491fbf4188450158',
#     'SM': 'a220a71737234036a6bfb0618f3e'
# }


# # print(f"⬆️ Uploading data to Firestore...")

# # Clean column names in deposit_df and index_df by stripping leading/trailing spaces
# deposit_df.columns = deposit_df.columns.str.strip()
# index_df.columns = index_df.columns.str.strip()

# # Ensure 'SCORECARD DATE' is datetime objects in both dataframes for consistent merging and processing
# # Use errors='coerce' to turn unparseable dates into NaT (Not a Time)
# deposit_df['SCORECARD DATE'] = pd.to_datetime(deposit_df['SCORECARD DATE'], errors='coerce').dt.date
# index_df['SCORECARD DATE'] = pd.to_datetime(index_df['SCORECARD DATE'], errors='coerce').dt.date


# # Drop rows with invalid dates if necessary (or handle them in the loop)
# deposit_df.dropna(subset=['SCORECARD DATE'], inplace=True)
# index_df.dropna(subset=['SCORECARD DATE'], inplace=True)


# # Add a 'userId' column to both dataframes based on the mapping
# deposit_df['userId'] = deposit_df['PLAYER'].map(user_id_map)
# index_df['userId'] = index_df['PLAYER'].map(user_id_map)

# # Drop rows where userId could not be mapped
# deposit_df.dropna(subset=['userId'], inplace=True)
# index_df.dropna(subset=['userId'], inplace=True)


# # Merge index_df and deposit_df to get the SCORE from index_df for correct round ID matching and data inclusion
# merged_hole_data = pd.merge(
#     deposit_df,
#     index_df[['PLAYER', 'CC', 'SCORECARD DATE', 'ROUNDS', 'SCORE', 'userId']],
#     on=['PLAYER', 'CC', 'SCORECARD DATE', 'ROUNDS', 'userId'],
#     how='left',
#     suffixes=('_hole', '_scorecard') # Use suffixes to differentiate score/rounds columns
# )

# # Drop rows where the merge failed to find a matching round in index_df (SCORE_scorecard will be NaN)
# merged_hole_data.dropna(subset=['SCORE_scorecard'], inplace=True)


# # Group by userId and the components that form the round document ID
# # Grouping by userId, CC, Date, Rounds (from deposit_df), and Score (from index_df)
# round_groups = merged_hole_data.groupby(['userId', 'CC', 'SCORECARD DATE', 'ROUNDS', 'SCORE_scorecard'])


# # print(f"Iterating through {len(round_groups)} unique rounds to upload data...")

# # for (userId, cc, date, rounds_hole, score_scorecard), hole_data_group in round_groups:
# #     # Generate the round document ID based on the structure: {Course}_{Date}_{RoundNumber}_{ScoreFromIndexDF}
# #     date_str = date.strftime('%Y-%m-%d') if pd.notna(date) else "N_A_date"
# #     rounds_str = str(int(rounds_hole)) if pd.notna(rounds_hole) and pd.api.types.is_integer(rounds_hole) else str(rounds_hole)
# #     score_str = str(int(score_scorecard)) if pd.notna(score_scorecard) and pd.api.types.is_integer(score_scorecard) else str(score_scorecard)

# #     round_id = f"{cc}_{date_str}_{rounds_str}_{score_str}"

# #     # Construct the path for the round document under the user's rounds collection
# #     round_doc_path = f"players/{userId}/rounds/{round_id}"


# #     # Get a reference to the round document
# #     round_ref = db.document(round_doc_path)

# #     # Aggregate summary data for the round from index_df
# #     # Find the corresponding row in index_df using the grouping keys
# #     round_summary_data_row = index_df[
# #         (index_df['userId'] == userId) &
# #         (index_df['CC'] == cc) &
# #         (index_df['SCORECARD DATE'] == date) &
# #         (index_df['ROUNDS'] == rounds_hole) &
# #         (index_df['SCORE'] == score_scorecard)
# #     ]

# #     round_summary_data = {}
# #     if not round_summary_data_row.empty:
# #         # Take the first row if multiple match (should ideally be unique)
# #         summary_row = round_summary_data_row.iloc[0]
# #         # Prepare summary data dictionary (excluding columns that will be in hole data or are redundant)
# #         round_summary_data = summary_row.drop(columns=[
# #             'PLAYER', 'CC', 'SCORECARD DATE', 'ROUNDS', 'SCORE', 'userId'
# #         ], errors='ignore').to_dict()
# #         round_summary_data['total_score'] = int(summary_row['SCORE']) # Add total score explicitly
# #         round_summary_data['date'] = date_str # Add date as string
# #         round_summary_data['course'] = cc # Add course explicitly
# #         round_summary_data['round_number'] = rounds_str # Add round number explicitly


# #     # Set the round document in Firestore with error handling
# #     try:
# #         round_ref.set(round_summary_data, merge=True) # Use merge=True to avoid overwriting existing data
# #         print(f"\n  Successfully set round document: {round_doc_path}")

# #         # Get a reference to the 'holes' subcollection under this round document
# #         holes_subcollection_ref = round_ref.collection("holes")

# #         # Iterate through each row (each hole) within the current round's group DataFrame
# #         print(f"\n  Uploading {len(hole_data_group)} holes for round: {round_id}")
# #         for _, hole_row in hole_data_group.iterrows():
# #             hole_num = hole_row.get("HOLE") # Hole number from deposit_df (merged)
# #             if pd.isna(hole_num):
# #                 print(f"    Skipping hole data for a row in round {round_id} due to missing HOLE number.")
# #                 continue # Skip if hole number is missing

# #             hole_num_str = str(int(hole_num)) if pd.api.types.is_integer(hole_num) else str(hole_num)

# #             # Prepare data for the hole document
# #             # Include relevant columns from the merged row (hole data)
# #             hole_data = hole_row.drop(columns=[
# #                 'ROUNDS', 'SCORECARD DATE', 'CC', 'PLAYER', 'userId', # Exclude grouping/parent key columns from hole data
# #                 'SCORE_hole', 'SCORE_scorecard' # Exclude other potentially redundant columns
# #             ], errors='ignore').to_dict()

# #             # Convert potential numpy types to native Python types and handle NaN/None
# #             cleaned_hole_data = {}
# #             for key, value in hole_data.items():
# #                 if pd.isna(value):
# #                     cleaned_hole_data[key] = None # Store NaN as None
# #                 # Convert datetime.date/Timestamp objects to string explicitly
# #                 elif isinstance(value, (datetime, pd.Timestamp)):
# #                     cleaned_hole_data[key] = value.strftime('%Y-%m-%d')
# #                 elif isinstance(value, (np.generic, np.ndarray)): # Handle NumPy types
# #                     # Convert NumPy types to native Python types (scalar or list)
# #                     cleaned_hole_data[key] = value.item() if value.size == 1 else value.tolist()
# #                 elif isinstance(value, pd.Series): # Handle potential Series objects
# #                      # Assuming the Series only contains one value or we only need the first
# #                     cleaned_hole_data[key] = value.iloc[0] if not value.empty and not pd.isna(value.iloc[0]) else None
# #                 else:
# #                     cleaned_hole_data[key] = value # Keep other types as is

# #             # Reference to the hole document in the 'holes' subcollection
# #             hole_doc_ref = holes_subcollection_ref.document(hole_num_str)

# #             # Set the data in Firestore
# #             try:
# #                 hole_doc_ref.set(cleaned_hole_data)
# #             except Exception as e:
# #                 print(f"    ❌ Error uploading hole {hole_num_str} for round {round_id}: {e}")

# #     except Exception as e:
# #         print(f"  ❌ Error accessing/processing round document {round_id}: {e}")


# # print(f"\n✅ Upload of data to 'players/{{userId}}/rounds/{{roundId}}' and 'players/{{userId}}/rounds/{{roundId}}/holes/{{holeNumber}}' complete!")

✅ Firestore client obtained successfully.
✅ deposit_df loaded successfully.
✅ index_df loaded successfully.


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
print(merged_hole_data.columns)

Index(['ROUNDS', 'SCORECARD DATE', 'CC', 'TEE', 'HOLE', 'YDS', 'HDCP', 'PAR',
       'SCORE_hole', 'PUTT', 'DD', 'DDTYPE', 'PINDIST', '1STPU', 'PLAYER',
       'ESC', 'SCORE_scorecard'],
      dtype='object')


# Task
Upload the data from "/content/drive/MyDrive/MASTER-FIREBASE.xlsx" to Firestore, structuring it under the path "/users/{userId}/course_summaries/{courseId}" for course summaries and "/users/{userId}/course_summaries/{courseId}/holes/{holeNumber}" for hole-by-hole data, using the service account key from "/content/drive/MyDrive/serviceAccountKey.json".

## Identify user and course

### Subtask:
Determine how to extract the `userId` and `courseId` from your data or define them based on your application's logic.


**Reasoning**:
Examine the columns of `deposit_df` and `index_df` to identify potential `userId` and `courseId` based on the available data. The previous step already loaded the data and cleaned some columns. The `merged_hole_data` dataframe contains information from both original dataframes.



In [ ]:
print("Columns in deposit_df:")
print(deposit_df.columns)
print("\nColumns in index_df:")
print(index_df.columns)
print("\nSample data from deposit_df (first 5 rows):")
display(deposit_df.head())
print("\nSample data from index_df (first 5 rows):")
display(index_df.head())

Columns in deposit_df:
Index(['ROUNDS', 'SCORECARD DATE', 'CC', 'TEE', 'HOLE', 'YDS', 'HDCP', 'PAR',
       'SCORE', 'PUTT', 'DD', 'DDTYPE', 'PINDIST', '1STPU', 'PLAYER', 'ESC'],
      dtype='object')

Columns in index_df:
Index(['ROUNDS', 'SCORECARD DATE', 'MES', 'YEAR', 'CC', 'PLAYER', 'SCORE',
       'ESC', 'PUTT', 'DD', 'FIR', 'PINDIST', '1STPU', 'GIR', 'SCRAM', '1PUTT',
       '2PUTT', '3PUTT', '4PUTT', '1STPUTT', '1STPUTT3', '1STPUTT4',
       '1STPUTT5', 'GIR2', 'GIR3', 'GIR4', 'GIR5', 'BIRDIE3', 'BIRDIE4',
       'BIRDIE5', 'PAR3', 'PAR4', 'PAR5', 'BOGEY3', 'BOGEY4', 'BOGEY5',
       'DOBLE3', 'DOBLE4', 'DOBLE5', 'TRIPLEOR3', 'TRIPLEOR4', 'TRIPLEOR5',
       'DDMAX', 'DDMAXFIR', 'IDA', 'VUELTA', 'MEJOR', 'DIF+/-'],
      dtype='object')

Sample data from deposit_df (first 5 rows):


ROUNDS SCORECARD DATE      CC     TEE  HOLE  YDS  HDCP  PAR  SCORE  PUTT  \
0       1     2023-04-07  GUAY#2  BLANCA     1  554     6    5      4     1   
1       1     2023-04-07  GUAY#2  BLANCA     2  427     2    4      4     1   
2       1     2023-04-07  GUAY#2  BLANCA     3  370    14    4      5     2   
3       1     2023-04-07  GUAY#2  BLANCA     4  167    16    3      4     1   
4       1     2023-04-07  GUAY#2  BLANCA     5  397     4    4      8     3   

   DD DDTYPE  PINDIST  1STPU PLAYER  ESC  
0 NaN    NaN    185.0    NaN     ML    4  
1 NaN    NaN    142.0    NaN     ML    4  
2 NaN    NaN    123.0    NaN     ML    5  
3 NaN    NaN      NaN    NaN     ML    4  
4 NaN    NaN     79.0    NaN     ML    6


Sample data from index_df (first 5 rows):


ROUNDS SCORECARD DATE      MES  YEAR          CC PLAYER  SCORE  ESC  PUTT  \
0       1     2023-04-07   ABRIL   2023      GUAY#2     ML     86   84    31   
1       2     2023-04-07   ABRIL   2023      GUAY#2     SM     87   86    33   
2       3     2023-05-03    MAYO   2023  FUNDADORES     ML    100   93    35   
3       4     2023-05-03    MAYO   2023  FUNDADORES     SM     95   92    32   
4       5     2023-05-17    MAYO   2023  FUNDADORES     ML     97   94    36   

   DD  ...  DOBLE5  TRIPLEOR3  TRIPLEOR4  TRIPLEOR5  DDMAX  DDMAXFIR  IDA  \
0 NaN  ...       0          0          1          0      0      -      44   
1 NaN  ...       0          0          0          1      0      -      47   
2 NaN  ...       1          1          3          1      0      -      43   
3 NaN  ...       1          1          2          0      0      -      44   
4 NaN  ...       2          1          1          0      0      -      47   

   VUELTA   MEJOR  DIF+/-  
0      42  Vuelta      -2  
1      40  Vuelta      -7  
2      57     Ida      14  
3      51     Ida       7  
4      50     Ida       3  

[5 rows x 48 columns]

## Modify data grouping

### Subtask:
Adjust the code to group the `merged_hole_data` by user and course.


## Construct firestore paths

### Subtask:
Update the code to construct the Firestore document paths based on the new structure: `/users/{userId}/course_summaries/{courseId}` for the course summary document and `/users/{userId}/course_summaries/{courseId}/holes/{holeNumber}` for the hole documents.


**Reasoning**:
Modify the existing code to construct the new Firestore document paths based on the user, course, and hole number, and update the Firestore set operations to use these new paths.



In [ ]:
from firebase_admin import firestore
from datetime import datetime
import numpy as np
import pandas as pd

# Assuming deposit_df and index_df are already loaded from Google Drive
# Assuming the Firebase Admin SDK is already initialized and 'db' is the Firestore client

if 'deposit_df' not in locals() or deposit_df.empty:
    print("❌ deposit_df not found or is empty. Cannot proceed with data processing and upload.")
elif 'index_df' not in locals() or index_df.empty:
    print("❌ index_df not found or is empty. Cannot proceed with data processing and upload.")
else:
    print("✅ DataFrames loaded and ready for processing.")

    # Clean column names in deposit_df and index_df by stripping leading/trailing spaces
    deposit_df.columns = deposit_df.columns.str.strip()
    index_df.columns = index_df.columns.str.strip()

    # Ensure 'SCORECARD DATE' is datetime objects in both dataframes for consistent merging
    # Use errors='coerce' to turn unparseable dates into NaT (Not a Time)
    deposit_df['SCORECARD DATE'] = pd.to_datetime(deposit_df['SCORECARD DATE'], errors='coerce').dt.date
    index_df['SCORECARD DATE'] = pd.to_datetime(index_df['SCORECARD DATE'], errors='coerce').dt.date

    # Drop rows with invalid dates if necessary (or handle them in the loop)
    deposit_df.dropna(subset=['SCORECARD DATE'], inplace=True)
    index_df.dropna(subset=['SCORECARD DATE'], inplace=True)

    # Merge index_df and deposit_df
    # We merge on Player, Course, Scorecard Date, and Rounds to link hole data to specific rounds
    merged_hole_data = pd.merge(
        deposit_df,
        index_df[['PLAYER', 'CC', 'SCORECARD DATE', 'ROUNDS', 'SCORE']],
        on=['PLAYER', 'CC', 'SCORECARD DATE', 'ROUNDS'],
        how='left',
        suffixes=('_hole', '_scorecard') # Use suffixes to differentiate score columns if needed
    )

    # Drop rows where the merge failed (i.e., hole data without a matching scorecard entry)
    merged_hole_data.dropna(subset=['SCORE_scorecard'], inplace=True)

    print("✅ Data merged successfully.")
    print("\nSample of merged data (first 5 rows):")
    display(merged_hole_data.head())

    # Group by the components that will define the course summary document (User and Course)
    user_course_groups = merged_hole_data.groupby(['PLAYER', 'CC'])

    print(f"\nReady to process {len(user_course_groups)} unique user-course combinations for upload.")

    # --- Firestore Upload Logic ---
    # This part will iterate through the grouped data and construct Firestore paths

    print(f"Iterating through {len(user_course_groups)} unique user-course combinations to construct paths...")

    for (player, cc), user_course_group in user_course_groups:
        # Construct the path for the course summary document
        # Using 'PLAYER' as userId and 'CC' as courseId based on data analysis
        course_summary_path = f"users/{player}/course_summaries/{cc}"
        print(f"\n  Constructed course summary path: {course_summary_path}")

        # The base path for the holes subcollection is derived from the course summary path
        holes_base_path = f"{course_summary_path}/holes"
        print(f"  Constructed holes subcollection base path: {holes_base_path}")

        # Note: The actual Firestore set operations will be done in subsequent steps
        # This loop is currently just demonstrating path construction based on grouping.


    print("\nPath construction complete. Ready for data upload in the next steps.")

✅ DataFrames loaded and ready for processing.
✅ Data merged successfully.

Sample of merged data (first 5 rows):


ROUNDS SCORECARD DATE      CC     TEE  HOLE  YDS  HDCP  PAR  SCORE_hole  \
0       1     2023-04-07  GUAY#2  BLANCA     1  554     6    5           4   
1       1     2023-04-07  GUAY#2  BLANCA     2  427     2    4           4   
2       1     2023-04-07  GUAY#2  BLANCA     3  370    14    4           5   
3       1     2023-04-07  GUAY#2  BLANCA     4  167    16    3           4   
4       1     2023-04-07  GUAY#2  BLANCA     5  397     4    4           8   

   PUTT  DD DDTYPE  PINDIST  1STPU PLAYER  ESC  SCORE_scorecard  
0     1 NaN    NaN    185.0    NaN     ML    4               86  
1     1 NaN    NaN    142.0    NaN     ML    4               86  
2     2 NaN    NaN    123.0    NaN     ML    5               86  
3     1 NaN    NaN      NaN    NaN     ML    4               86  
4     3 NaN    NaN     79.0    NaN     ML    6               86


Ready to process 44 unique user-course combinations for upload.
Iterating through 44 unique user-course combinations to construct paths...

  Constructed course summary path: users/ML/course_summaries/ARRAYANES
  Constructed holes subcollection base path: users/ML/course_summaries/ARRAYANES/holes

  Constructed course summary path: users/ML/course_summaries/BRICEÑO
  Constructed holes subcollection base path: users/ML/course_summaries/BRICEÑO/holes

  Constructed course summary path: users/ML/course_summaries/BRITANIA
  Constructed holes subcollection base path: users/ML/course_summaries/BRITANIA/holes

  Constructed course summary path: users/ML/course_summaries/CARMEL
  Constructed holes subcollection base path: users/ML/course_summaries/CARMEL/holes

  Constructed course summary path: users/ML/course_summaries/CORAL SPRINGS
  Constructed holes subcollection base path: users/ML/course_summaries/CORAL SPRINGS/holes

  Constructed course summary path: users/ML/course_summaries/CRANDON

## Upload course summary data

### Subtask:
Upload the relevant summary data for each user and course to the `/users/{userId}/course_summaries/{courseId}` document.


**Reasoning**:
Populate the course_summary_data dictionary with relevant summary information from the index_df, and include error handling for the Firestore set operation.



In [ ]:
from firebase_admin import firestore

if 'index_df' not in locals() or index_df.empty:
    print("❌ index_df not found or is empty. Cannot proceed with data upload.")
else:
    # Clean column names in index_df by stripping leading/trailing spaces
    index_df.columns = index_df.columns.str.strip()

    # Ensure 'SCORECARD DATE' is datetime objects in both dataframes for consistent merging
    # Use errors='coerce' to turn unparseable dates into NaT (Not a Time)
    index_df['SCORECARD DATE'] = pd.to_datetime(index_df['SCORECARD DATE'], errors='coerce').dt.date
    deposit_df['SCORECARD DATE'] = pd.to_datetime(deposit_df['SCORECARD DATE'], errors='coerce').dt.date

    # Drop rows with invalid dates if necessary (or handle them in the loop)
    index_df.dropna(subset=['SCORECARD DATE'], inplace=True)
    deposit_df.dropna(subset=['SCORECARD DATE'], inplace=True)

    # Merge index_df and deposit_df to get the SCORE from index_df for correct round ID matching
    merged_hole_data = pd.merge(
        deposit_df,
        index_df[['PLAYER', 'CC', 'SCORECARD DATE', 'ROUNDS', 'SCORE']],
        on=['PLAYER', 'CC', 'SCORECARD DATE', 'ROUNDS'],
        how='left',
        suffixes=('_hole', '_scorecard') # Use suffixes to differentiate score/rounds columns
    )

    # Group by the components that form the round document key components (User and Course)
    user_course_groups = merged_hole_data.groupby(['PLAYER', 'CC'])

    print(f"Iterating through {len(user_course_groups)} unique user-course combinations to upload data...")

    for (player, cc), user_course_group in user_course_groups:
        # Construct the path for the course summary document
        course_summary_path = f"users/{player}/course_summaries/{cc}"

        # Get a reference to the course summary document
        course_summary_ref = db.document(course_summary_path)

        # Aggregate summary data from index_df for the current player and course
        player_course_index_df = index_df[(index_df['PLAYER'] == player) & (index_df['CC'] == cc)]

        # Prepare data for the course summary document
        course_summary_data = {
            "player": player,
            "course": cc,
            "number_of_rounds": len(player_course_index_df),
            "average_score": player_course_index_df['SCORE'].mean() if not player_course_index_df.empty else None,
            "last_updated": firestore.SERVER_TIMESTAMP # Example field
        }

        # Set the course summary document in Firestore with error handling
        try:
            course_summary_ref.set(course_summary_data, merge=True) # Use merge=True to avoid overwriting existing data
            print(f"\n  Successfully set course summary document: {course_summary_path}")

            # Get a reference to the 'holes' subcollection under this course summary document
            holes_subcollection_ref = course_summary_ref.collection("holes")

            # Iterate through each row (each hole) within the current user-course group DataFrame
            print(f"\n  Uploading {len(user_course_group)} holes for user: {player}, course: {cc}")
            for _, hole_row in user_course_group.iterrows():
                hole_num = hole_row.get("HOLE") # Hole number from deposit_df (merged)
                if pd.isna(hole_num):
                    print(f"    Skipping hole data for a row in user {player}, course {cc} due to missing HOLE number.")
                    continue # Skip if hole number is missing

                hole_num_str = str(int(hole_num)) if pd.api.types.is_integer(hole_num) else str(hole_num)

                # Prepare data for the hole document
                # Include relevant columns from the merged row (hole data)
                hole_data = hole_row.drop(columns=[
                    'PLAYER', 'CC', # Exclude grouping/parent key columns from hole data
                    'SCORECARD DATE', 'ROUNDS', 'SCORE_hole', 'SCORE_scorecard' # Exclude other potentially redundant columns
                ], errors='ignore').to_dict()

                # Convert potential numpy types to native Python types and handle NaN/None
                cleaned_hole_data = {}
                for key, value in hole_data.items():
                    if pd.isna(value):
                        cleaned_hole_data[key] = None # Store NaN as None
                    # Convert datetime.date/Timestamp objects to string explicitly
                    elif isinstance(value, (datetime, pd.Timestamp)):
                        cleaned_hole_data[key] = value.strftime('%Y-%m-%d')
                    elif isinstance(value, (np.generic, np.ndarray)): # Handle NumPy types
                        # Convert NumPy types to native Python types (scalar or list)
                        cleaned_hole_data[key] = value.item() if value.size == 1 else value.tolist()
                    elif isinstance(value, pd.Series): # Handle potential Series objects
                        cleaned_hole_data[key] = value.iloc[0] if not pd.isna(value.iloc[0]) else None
                    else:
                        cleaned_hole_data[key] = value # Keep other types as is

                # Reference to the hole document in the 'holes' subcollection
                hole_doc_ref = holes_subcollection_ref.document(hole_num_str)

                # Set the data in Firestore
                try:
                    hole_doc_ref.set(cleaned_hole_data)
                except Exception as e:
                    print(f"    ❌ Error uploading hole {hole_num_str} for user {player}, course {cc}: {e}")

        except Exception as e:
            print(f"  ❌ Error accessing/processing user {player}, course {cc}: {e}")

print(f"\n✅ Upload of data to 'users/{{userId}}/course_summaries/{{courseId}}' and 'users/{{userId}}/course_summaries/{{courseId}}/holes/{{holeNumber}}' complete!")


Iterating through 44 unique user-course combinations to upload data...
  ❌ Error accessing/processing user ML, course ARRAYANES: 403 Permission denied on resource project golf-scorecard-1847c. [reason: "CONSUMER_INVALID"
domain: "googleapis.com"
metadata {
  key: "service"
  value: "firestore.googleapis.com"
}
metadata {
  key: "containerInfo"
  value: "golf-scorecard-1847c"
}
metadata {
  key: "consumer"
  value: "projects/golf-scorecard-1847c"
}
, locale: "en-US"
message: "Permission denied on resource project golf-scorecard-1847c."
, links {
  description: "Google developers console"
  url: "https://console.developers.google.com"
}
]
  ❌ Error accessing/processing user ML, course BRICEÑO: 403 Permission denied on resource project golf-scorecard-1847c. [reason: "CONSUMER_INVALID"
domain: "googleapis.com"
metadata {
  key: "service"
  value: "firestore.googleapis.com"
}
metadata {
  key: "containerInfo"
  value: "golf-scorecard-1847c"
}
metadata {
  key: "consumer"
  value: "projects/

## Upload hole data to subcollection

### Subtask:
Iterate through the grouped hole data and upload each hole's information to the 'holes' subcollection under the corresponding course summary document.


**Reasoning**:
Iterate through the grouped hole data and upload each hole's information to the 'holes' subcollection under the corresponding course summary document. This involves extracting hole data, preparing it for Firestore, and setting the document within the subcollection with error handling.



In [ ]:
from firebase_admin import firestore
from datetime import datetime
import numpy as np
import pandas as pd

if 'index_df' not in locals() or index_df.empty:
    print("❌ index_df not found or is empty. Cannot proceed with data upload.")
else:
    # Clean column names in index_df by stripping leading/trailing spaces
    index_df.columns = index_df.columns.str.strip()

    # Ensure 'SCORECARD DATE' is datetime objects in both dataframes for consistent merging
    # Use errors='coerce' to turn unparseable dates into NaT (Not a Time)
    index_df['SCORECARD DATE'] = pd.to_datetime(index_df['SCORECARD DATE'], errors='coerce').dt.date
    deposit_df['SCORECARD DATE'] = pd.to_datetime(deposit_df['SCORECARD DATE'], errors='coerce').dt.date

    # Drop rows with invalid dates if necessary (or handle them in the loop)
    index_df.dropna(subset=['SCORECARD DATE'], inplace=True)
    deposit_df.dropna(subset=['SCORECARD DATE'], inplace=True)

    # Merge index_df and deposit_df to get the SCORE from index_df for correct round ID matching
    merged_hole_data = pd.merge(
        deposit_df,
        index_df[['PLAYER', 'CC', 'SCORECARD DATE', 'ROUNDS', 'SCORE']],
        on=['PLAYER', 'CC', 'SCORECARD DATE', 'ROUNDS'],
        how='left',
        suffixes=('_hole', '_scorecard') # Use suffixes to differentiate score/rounds columns
    )

    # Group by the components that form the round document key components (User and Course)
    user_course_groups = merged_hole_data.groupby(['PLAYER', 'CC'])

    print(f"Iterating through {len(user_course_groups)} unique user-course combinations to upload data...")

    for (player, cc), user_course_group in user_course_groups:
        # Construct the path for the course summary document
        course_summary_path = f"users/{player}/course_summaries/{cc}"

        # Get a reference to the course summary document
        course_summary_ref = db.document(course_summary_path)

        # Aggregate summary data from index_df for the current player and course
        player_course_index_df = index_df[(index_df['PLAYER'] == player) & (index_df['CC'] == cc)]

        # Prepare data for the course summary document
        course_summary_data = {
            "player": player,
            "course": cc,
            "number_of_rounds": len(player_course_index_df),
            "average_score": player_course_index_df['SCORE'].mean() if not player_course_index_df.empty else None,
            "last_updated": firestore.SERVER_TIMESTAMP # Example field
        }

        # Set the course summary document in Firestore with error handling
        try:
            course_summary_ref.set(course_summary_data, merge=True) # Use merge=True to avoid overwriting existing data
            print(f"\n  Successfully set course summary document: {course_summary_path}")

            # Get a reference to the 'holes' subcollection under this course summary document
            holes_subcollection_ref = course_summary_ref.collection("holes")

            # Iterate through each row (each hole) within the current user-course group DataFrame
            print(f"\n  Uploading {len(user_course_group)} holes for user: {player}, course: {cc}")
            for _, hole_row in user_course_group.iterrows():
                hole_num = hole_row.get("HOLE") # Hole number from deposit_df (merged)
                if pd.isna(hole_num):
                    print(f"    Skipping hole data for a row in user {player}, course {cc} due to missing HOLE number.")
                    continue # Skip if hole number is missing

                hole_num_str = str(int(hole_num)) if pd.api.types.is_integer(hole_num) else str(hole_num)

                # Prepare data for the hole document
                # Include relevant columns from the merged row (hole data)
                hole_data = hole_row.drop(columns=[
                    'PLAYER', 'CC', # Exclude grouping/parent key columns from hole data
                    'SCORECARD DATE', 'ROUNDS', 'SCORE_hole', 'SCORE_scorecard' # Exclude other potentially redundant columns
                ], errors='ignore').to_dict()

                # Convert potential numpy types to native Python types and handle NaN/None
                cleaned_hole_data = {}
                for key, value in hole_data.items():
                    if pd.isna(value):
                        cleaned_hole_data[key] = None # Store NaN as None
                    # Convert datetime.date/Timestamp objects to string explicitly
                    elif isinstance(value, (datetime, pd.Timestamp)):
                        cleaned_hole_data[key] = value.strftime('%Y-%m-%d')
                    elif isinstance(value, (np.generic, np.ndarray)): # Handle NumPy types
                        # Convert NumPy types to native Python types (scalar or list)
                        cleaned_hole_data[key] = value.item() if value.size == 1 else value.tolist()
                    elif isinstance(value, pd.Series): # Handle potential Series objects
                         # Assuming the Series only contains one value or we only need the first
                        cleaned_hole_data[key] = value.iloc[0] if not value.empty and not pd.isna(value.iloc[0]) else None
                    else:
                        cleaned_hole_data[key] = value # Keep other types as is

                # Reference to the hole document in the 'holes' subcollection
                hole_doc_ref = holes_subcollection_ref.document(hole_num_str)

                # Set the data in Firestore
                try:
                    hole_doc_ref.set(cleaned_hole_data)
                except Exception as e:
                    print(f"    ❌ Error uploading hole {hole_num_str} for user {player}, course {cc}: {e}")

        except Exception as e:
            print(f"  ❌ Error accessing/processing user {player}, course {cc}: {e}")

print(f"\n✅ Upload of data to 'users/{{userId}}/course_summaries/{{courseId}}' and 'users/{{userId}}/course_summaries/{{courseId}}/holes/{{holeNumber}}' complete!")

Iterating through 44 unique user-course combinations to upload data...
  ❌ Error accessing/processing user ML, course ARRAYANES: 403 Permission denied on resource project golf-scorecard-1847c. [reason: "CONSUMER_INVALID"
domain: "googleapis.com"
metadata {
  key: "service"
  value: "firestore.googleapis.com"
}
metadata {
  key: "containerInfo"
  value: "golf-scorecard-1847c"
}
metadata {
  key: "consumer"
  value: "projects/golf-scorecard-1847c"
}
, locale: "en-US"
message: "Permission denied on resource project golf-scorecard-1847c."
, links {
  description: "Google developers console"
  url: "https://console.developers.google.com"
}
]
  ❌ Error accessing/processing user ML, course BRICEÑO: 403 Permission denied on resource project golf-scorecard-1847c. [reason: "CONSUMER_INVALID"
domain: "googleapis.com"
metadata {
  key: "service"
  value: "firestore.googleapis.com"
}
metadata {
  key: "containerInfo"
  value: "golf-scorecard-1847c"
}
metadata {
  key: "consumer"
  value: "projects/

## Review and refine

### Subtask:
Review the data structure in Firestore after the attempted upload to understand the current state and identify discrepancies based on the desired structure, given the permission errors encountered.


## Review and refine retry 1

### Subtask:
Review the data structure in Firestore after the attempted upload to understand the current state and identify discrepancies based on the desired structure. Given the previous permission errors, focus on verifying if any documents were created at all, or if the error prevented any writes. This step requires manual review via the Firebase console.


## Summary:

### Data Analysis Key Findings

*   The process successfully identified 'PLAYER' as the `userId` and 'CC' as the `courseId` from the input dataframes (`deposit_df` and `index_df`) for structuring the Firestore data.
*   The code was correctly modified to group the `merged_hole_data` by user and course, aligning with the intended data organization for Firestore.
*   The code was updated to construct the correct Firestore paths: `/users/{userId}/course_summaries/{courseId}` for course summaries and `/users/{userId}/course_summaries/{courseId}/holes/{holeNumber}` for hole-by-hole data.
*   The code implemented the logic to aggregate course summary data (number of rounds and average score) and attempted to upload it to the corresponding Firestore document path.
*   The code implemented the logic to iterate through grouped hole data, format it, and attempted to upload each hole's data to the 'holes' subcollection under the course summary document.
*   During the attempted Firestore uploads, repeated "403 Permission denied" errors were encountered, indicating an issue with the service account's permissions to write to the Firestore project.

### Insights or Next Steps

*   The primary blocker for successful data upload is the "403 Permission denied" error. The next crucial step is to verify and update the service account permissions in the Google Cloud Platform console to ensure it has the necessary roles (e.g., Cloud Datastore User or Owner) to write data to Firestore.
*   After resolving the permission issues, re-execute the data upload script to confirm that data is successfully written to Firestore with the desired structure. A manual review of the Firestore database via the Firebase console will be necessary to confirm the successful upload and correct data structure.


# Task
Provide a step-by-step guide on how to integrate Google Drive, GitHub, Colab, and Firebase for data uploading, backup, and live server functions, focusing on a "practice mode" setup.

## Set up google cloud project and firebase

### Subtask:
Ensure you have a Google Cloud Project and a Firebase project set up and linked. This is where your Firestore database will reside.


In [ ]:
# Install the Firebase Admin SDK
!pip install firebase-admin

In [ ]:
import firebase_admin
from firebase_admin import credentials, firestore
import os

# Replace with the actual path to your service account key JSON file in Google Drive
# Make sure your Google Drive is mounted in Colab.
service_account_key_path = "/content/drive/MyDrive/serviceAccountKey.json"

# Initialize Firebase Admin SDK if it hasn't been initialized yet
try:
    # Check if a default app is already initialized to avoid re-initialization errors
    firebase_admin.get_app()
    print("Firebase Admin SDK already initialized.")
except ValueError:
    # If not initialized, create credentials and initialize the app
    if not os.path.exists(service_account_key_path):
        print(f"❌ Error: Service account key not found at '{service_account_key_path}'. Please check the path and ensure your Google Drive is mounted.")
    else:
        try:
            cred = credentials.Certificate(service_account_key_path)
            firebase_admin.initialize_app(cred)
            print("✅ Firebase Admin SDK initialized successfully.")
        except Exception as e:
            print(f"❌ Error initializing Firebase Admin SDK: {e}")

# Get the Firestore client
try:
    db = firestore.client()
    print("✅ Firestore client obtained successfully.")
except Exception as e:
    print(f"❌ Error getting Firestore client: {e}")

✅ Firebase Admin SDK initialized successfully.
✅ Firestore client obtained successfully.


## Configure Firebase Admin SDK

### Subtask:
Set up and configure the Firebase Admin SDK in your Colab environment using the generated service account key.

In [ ]:
import pandas as pd
from datetime import datetime
import numpy as np

# Assuming deposit_df and index_df are already loaded from Google Drive
# Assuming the Firebase Admin SDK is already initialized and 'db' is the Firestore client

if 'deposit_df' not in locals() or deposit_df.empty:
    print("❌ deposit_df not found or is empty. Cannot proceed with data processing and upload.")
elif 'index_df' not in locals() or index_df.empty:
    print("❌ index_df not found or is empty. Cannot proceed with data processing and upload.")
else:
    print("✅ DataFrames loaded and ready for processing.")

    # Clean column names in deposit_df and index_df by stripping leading/trailing spaces
    deposit_df.columns = deposit_df.columns.str.strip()
    index_df.columns = index_df.columns.str.strip()

    # Ensure 'SCORECARD DATE' is datetime objects in both dataframes for consistent merging
    # Use errors='coerce' to turn unparseable dates into NaT (Not a Time)
    deposit_df['SCORECARD DATE'] = pd.to_datetime(deposit_df['SCORECARD DATE'], errors='coerce').dt.date
    index_df['SCORECARD DATE'] = pd.to_datetime(index_df['SCORECARD DATE'], errors='coerce').dt.date

    # Drop rows with invalid dates if necessary (or handle them in the loop)
    deposit_df.dropna(subset=['SCORECARD DATE'], inplace=True)
    index_df.dropna(subset=['SCORECARD DATE'], inplace=True)

    # Merge index_df and deposit_df
    # We merge on Player, Course, Scorecard Date, and Rounds to link hole data to specific rounds
    merged_hole_data = pd.merge(
        deposit_df,
        index_df[['PLAYER', 'CC', 'SCORECARD DATE', 'ROUNDS', 'SCORE']],
        on=['PLAYER', 'CC', 'SCORECARD DATE', 'ROUNDS'],
        how='left',
        suffixes=('_hole', '_scorecard') # Use suffixes to differentiate score columns if needed
    )

    # Drop rows where the merge failed (i.e., hole data without a matching scorecard entry)
    merged_hole_data.dropna(subset=['SCORE_scorecard'], inplace=True)

    print("✅ Data merged successfully.")
    print("\nSample of merged data (first 5 rows):")
    display(merged_hole_data.head())

    # Group by the components that will define the course summary document (User and Course)
    user_course_groups = merged_hole_data.groupby(['PLAYER', 'CC'])

    print(f"\nReady to process {len(user_course_groups)} unique user-course combinations for upload.")

    # --- Firestore Upload Logic ---
    # This part will be implemented in the next steps to construct paths and upload data

    # for (player, cc), user_course_group in user_course_groups:
    #     # Construct the path for the course summary document
    #     course_summary_path = f"users/{player}/course_summaries/{cc}"

    #     # Get a reference to the course summary document
    #     course_summary_ref = db.document(course_summary_path)

    #     # Aggregate summary data from index_df for the current player and course
    #     player_course_index_df = index_df[(index_df['PLAYER'] == player) & (index_df['CC'] == cc)]

    #     # Prepare data for the course summary document
    #     course_summary_data = {
    #         "player": player,
    #         "course": cc,
    #         "number_of_rounds": len(player_course_index_df),
    #         "average_score": player_course_index_df['SCORE'].mean() if not player_course_index_df.empty else None,
    #         "last_updated": firestore.SERVER_TIMESTAMP # Example field
    #     }

    #     # Set the course summary document in Firestore with error handling
    #     try:
    #         course_summary_ref.set(course_summary_data, merge=True)
    #         print(f"\n  Successfully prepared course summary data for: {course_summary_path}")

    #         # Get a reference to the 'holes' subcollection under this course summary document
    #         holes_subcollection_ref = course_summary_ref.collection("holes")

    #         # Iterate through each row (each hole) within the current user-course group DataFrame
    #         print(f"  Preparing {len(user_course_group)} holes for user: {player}, course: {cc}")
    #         for _, hole_row in user_course_group.iterrows():
    #             hole_num = hole_row.get("HOLE")
    #             if pd.isna(hole_num):
    #                 print(f"    Skipping hole data for a row in user {player}, course {cc} due to missing HOLE number.")
    #                 continue

    #             hole_num_str = str(int(hole_num)) if pd.api.types.is_integer(hole_num) else str(hole_num)

    #             # Prepare data for the hole document
    #             hole_data = hole_row.drop(columns=[
    #                 'PLAYER', 'CC',
    #                 'SCORECARD DATE', 'ROUNDS', 'SCORE_hole', 'SCORE_scorecard'
    #             ], errors='ignore').to_dict()

    #             # Clean data for Firestore (handle NaN, NaT, numpy types)
    #             cleaned_hole_data = {}
    #             for key, value in hole_data.items():
    #                 if pd.isna(value):
    #                     cleaned_hole_data[key] = None
    #                 elif isinstance(value, (datetime, pd.Timestamp)):
    #                     cleaned_hole_data[key] = value.strftime('%Y-%m-%d')
    #                 elif isinstance(value, (np.generic, np.ndarray)):
    #                     cleaned_hole_data[key] = value.item() if value.size == 1 else value.tolist()
    #                 elif isinstance(value, pd.Series):
    #                      cleaned_hole_data[key] = value.iloc[0] if not value.empty and not pd.isna(value.iloc[0]) else None
    #                 else:
    #                     cleaned_hole_data[key] = value

    #             # Note: The actual upload of hole data will be in the next step to manage transaction/batch writes

    #     except Exception as e:
    #         print(f"  ❌ Error preparing data for user {player}, course {cc}: {e}")

    # print("\nData processing complete. Ready for Firestore upload in the next steps.")

❌ deposit_df not found or is empty. Cannot proceed with data processing and upload.


## Develop Colab notebooks for data processing and upload

### Subtask:
Create Colab notebooks to load data from Google Drive, process it as needed, and upload it to Firestore using the Firebase Admin SDK.

In [2]:
from firebase_admin import firestore
from datetime import datetime
import numpy as np
import pandas as pd

# Assuming deposit_df and index_df are already loaded and processed
# Assuming the Firebase Admin SDK is already initialized and 'db' is the Firestore client
# Assuming merged_hole_data and user_course_groups are already created from the previous steps

if 'deposit_df' not in locals() or deposit_df.empty:
    print("❌ deposit_df not found or is empty. Cannot proceed with data processing and upload.")
elif 'index_df' not in locals() or index_df.empty:
    print("❌ index_df not found or is empty. Cannot proceed with data processing and upload.")
elif 'db' not in locals():
    print("❌ Firestore client 'db' not found. Ensure Firebase Admin SDK is initialized.")
else:
    print("✅ DataFrames loaded and ready for processing.")

    # Clean column names in deposit_df and index_df by stripping leading/trailing spaces
    deposit_df.columns = deposit_df.columns.str.strip()
    index_df.columns = index_df.columns.str.strip()

    # Ensure 'SCORECARD DATE' is datetime objects in both dataframes for consistent merging and processing
    # Use errors='coerce' to turn unparseable dates into NaT (Not a Time)
    deposit_df['SCORECARD DATE'] = pd.to_datetime(deposit_df['SCORECARD DATE'], errors='coerce').dt.date
    index_df['SCORECARD DATE'] = pd.to_datetime(index_df['SCORECARD DATE'], errors='coerce').dt.date


    # Drop rows with invalid dates if necessary (or handle them in the loop)
    deposit_df.dropna(subset=['SCORECARD DATE'], inplace=True)
    index_df.dropna(subset=['SCORECARD DATE'], inplace=True)


    # Add a 'userId' column to both dataframes based on the mapping
    # Using the provided user IDs. Make sure 'PLAYER' column exists and matches keys.
    user_id_map = {
        'ML': '998b30f58115491fbf4188450158',
        'SM': 'a220a71737234036a6bfb0618f3e'
    }
    if 'PLAYER' in deposit_df.columns:
        deposit_df['userId'] = deposit_df['PLAYER'].map(user_id_map)
    else:
        print("❌ 'PLAYER' column not found in deposit_df. Cannot map user IDs.")
        # Handle this error appropriately, maybe skip upload or raise exception
        # For now, let's assume 'PLAYER' exists based on previous outputs.

    if 'PLAYER' in index_df.columns:
        index_df['userId'] = index_df['PLAYER'].map(user_id_map)
    else:
         print("❌ 'PLAYER' column not found in index_df. Cannot map user IDs.")
         # Handle this error

    # Drop rows where userId could not be mapped (if PLAYER names don't match map keys)
    deposit_df.dropna(subset=['userId'], inplace=True)
    index_df.dropna(subset=['userId'], inplace=True)


    # Merge index_df and deposit_df to get the SCORE from index_df for correct round ID matching and data inclusion
    # Merge on columns that uniquely identify a round
    merge_cols = ['PLAYER', 'CC', 'SCORECARD DATE', 'ROUNDS']
    if all(col in deposit_df.columns and col in index_df.columns for col in merge_cols):
        merged_hole_data = pd.merge(
            deposit_df,
            index_df[['PLAYER', 'CC', 'SCORECARD DATE', 'ROUNDS', 'SCORE', 'userId']],
            on=merge_cols + ['userId'], # Include userId in merge key
            how='left',
            suffixes=('_hole', '_scorecard') # Use suffixes to differentiate score/rounds columns
        )

        # Drop rows where the merge failed to find a matching round in index_df (SCORE_scorecard will be NaN)
        merged_hole_data.dropna(subset=['SCORE_scorecard'], inplace=True)
        print("✅ Data merged successfully.")
        print("\nSample of merged data (first 5 rows):")
        display(merged_hole_data.head())

    else:
        print("❌ Necessary columns for merging ('PLAYER', 'CC', 'SCORECARD DATE', 'ROUNDS') not found in both dataframes. Cannot proceed with merge.")
        merged_hole_data = pd.DataFrame() # Ensure merged_hole_data is empty


    # Group by userId and the components that form the round document ID
    # Grouping by userId, CC, Date, Rounds (from deposit_df), and Score (from index_df)
    if not merged_hole_data.empty:
        round_group_cols = ['userId', 'CC', 'SCORECARD DATE', 'ROUNDS_hole', 'SCORE_scorecard']
        # Ensure grouping columns exist before grouping
        if all(col in merged_hole_data.columns for col in round_group_cols):
             round_groups = merged_hole_data.groupby(round_group_cols)
             print(f"\nReady to process {len(round_groups)} unique rounds for upload.")

             # --- Firestore Upload Logic ---
             print(f"Iterating through {len(round_groups)} unique rounds to upload data...")

             for (userId, cc, date, rounds_hole, score_scorecard), hole_data_group in round_groups:
                 # Generate the round document ID based on the structure: {Course}_{Date}_{RoundNumber}_{ScoreFromIndexDF}
                 date_str = date.strftime('%Y-%m-%d') if pd.notna(date) else "N_A_date"
                 # Ensure rounds_hole and score_scorecard are treated as integers if possible for cleaner IDs
                 rounds_str = str(int(rounds_hole)) if pd.notna(rounds_hole) and pd.api.types.is_integer(rounds_hole) else str(rounds_hole)
                 score_str = str(int(score_scorecard)) if pd.notna(score_scorecard) and pd.api.types.is_integer(score_scorecard) else str(score_scorecard)


                 round_id = f"{cc}_{date_str}_{rounds_str}_{score_str}"

                 # Construct the path for the round document under the user's rounds collection
                 round_doc_path = f"players/{userId}/rounds/{round_id}"


                 # Get a reference to the round document
                 round_ref = db.document(round_doc_path)

                 # Aggregate summary data for the round from index_df
                 # Find the corresponding row in index_df using the grouping keys (adjusting for merged column names)
                 round_summary_data_row = index_df[
                     (index_df['userId'] == userId) &
                     (index_df['CC'] == cc) &
                     (index_df['SCORECARD DATE'] == date) &
                     (index_df['ROUNDS'] == rounds_hole) # Use original rounds column from index_df
                 ]

                 round_summary_data = {}
                 if not round_summary_data_row.empty:
                     # Take the first row if multiple match (should ideally be unique for a round)
                     summary_row = round_summary_data_row.iloc[0]
                     # Prepare summary data dictionary (excluding columns that will be in hole data or are redundant)
                     round_summary_data = summary_row.drop(columns=[
                         'PLAYER', 'CC', 'SCORECARD DATE', 'ROUNDS', 'userId'
                     ], errors='ignore').to_dict()
                     # Add explicitly desired fields for clarity in Firestore
                     round_summary_data['total_score'] = int(summary_row['SCORE']) if pd.notna(summary_row['SCORE']) else None
                     round_summary_data['date'] = date_str # Add date as string
                     round_summary_data['course'] = cc # Add course explicitly
                     round_summary_data['round_number'] = rounds_str # Add round number explicitly
                     round_summary_data['last_updated'] = firestore.SERVER_TIMESTAMP # Add a timestamp


                 # Set the round document in Firestore with error handling
                 try:
                     round_ref.set(round_summary_data, merge=True) # Use merge=True to avoid overwriting existing data
                     print(f"\n  Successfully set round document: {round_doc_path}")

                     # Get a reference to the 'holes' subcollection under this round document
                     holes_subcollection_ref = round_ref.collection("holes")

                     # Iterate through each row (each hole) within the current round's group DataFrame
                     print(f"  Uploading {len(hole_data_group)} holes for round: {round_id}")
                     for _, hole_row in hole_data_group.iterrows():
                         hole_num = hole_row.get("HOLE") # Hole number from deposit_df (merged)
                         if pd.isna(hole_num):
                             print(f"    Skipping hole data for a row in round {round_id} due to missing HOLE number.")
                             continue # Skip if hole number is missing

                         # Ensure hole_num is an integer before converting to string for document ID
                         try:
                             # Convert to integer first to handle potential float format, then to string
                             hole_num_str = str(int(float(hole_num))) if pd.notna(hole_num) else "N_A_hole"
                         except (ValueError, TypeError):
                              hole_num_str = str(hole_num) # Use original value if conversion to int/float fails


                         # Prepare data for the hole document
                         # Include relevant columns from the merged row (hole data)
                         hole_data = hole_row.drop(columns=[
                             'PLAYER', 'CC', 'userId', # Exclude grouping/parent key columns from hole data
                             'SCORECARD DATE', 'ROUNDS_hole', 'ROUNDS_scorecard', 'SCORE_hole', 'SCORE_scorecard' # Exclude other potentially redundant columns
                         ], errors='ignore').to_dict()

                         # Convert potential numpy types to native Python types and handle NaN/None
                         cleaned_hole_data = {}
                         for key, value in hole_data.items():
                             if pd.isna(value):
                                 cleaned_hole_data[key] = None # Store NaN as None
                             # Convert datetime.date/Timestamp objects to string explicitly
                             elif isinstance(value, (datetime, pd.Timestamp)):
                                 cleaned_hole_data[key] = value.strftime('%Y-%m-%d')
                             elif isinstance(value, (np.generic, np.ndarray)): # Handle NumPy types
                                 # Convert NumPy types to native Python types (scalar or list)
                                 cleaned_hole_data[key] = value.item() if value.size == 1 else value.tolist()
                             elif isinstance(value, pd.Series): # Handle potential Series objects
                                  cleaned_hole_data[key] = value.iloc[0] if not value.empty and not pd.isna(value.iloc[0]) else None
                             else:
                                 cleaned_hole_data[key] = value # Keep other types as is

                         # Reference to the hole document in the 'holes' subcollection
                         # Use the hole number as the document ID
                         hole_doc_ref = holes_subcollection_ref.document(hole_num_str)

                         # Set the data in Firestore
                         try:
                             hole_doc_ref.set(cleaned_hole_data)
                         except Exception as e:
                             print(f"    ❌ Error uploading hole {hole_num_str} for round {round_id}: {e}")

                 except Exception as e:
                     print(f"  ❌ Error accessing/processing round document {round_id}: {e}")

             print(f"\n✅ Attempted upload of data to 'players/{{userId}}/rounds/{{roundId}}' and 'players/{{userId}}/rounds/{{roundId}}/holes/{{holeNumber}}'. Please check your Firestore database to confirm.")

        else:
            print("❌ Necessary columns for grouping ('userId', 'CC', 'SCORECARD DATE', 'ROUNDS_hole', 'SCORE_scorecard') not found in merged_hole_data. Cannot proceed with grouping and upload.")

    else:
        print("❌ merged_hole_data is empty. Cannot proceed with grouping and upload.")

❌ deposit_df not found or is empty. Cannot proceed with data processing and upload.


# Task
Create a plan to develop an application where users can view performance charts updated in real-time upon scorecard uploads, and a backend system to generate HTML reports from the same database.

## Review firestore data structure

### Subtask:
Confirm the current Firestore data structure is suitable for both displaying individual user performance in an app and for generating aggregated backend reports. Adjust if necessary.


## Develop real-time data access for user app

### Subtask:
Plan and implement how your user-facing application will access and display the performance data from Firestore in near real-time. This might involve using Firestore's real-time listeners.


**Reasoning**:
Outline the key data points and aggregations needed from Firestore for performance charts and describe how real-time listeners will be used in the user application.



In [ ]:
# 1. Key Data Points and Aggregations for User-Facing Application
# The user-facing application will primarily need data to display performance charts
# for a specific user, potentially filtering by course.

# From users/{userId}/course_summaries/{courseId}:
# - course: Course name (for filtering/display)
# - number_of_rounds: Total rounds played on this course (for context)
# - average_score: Average score on this course (key performance indicator)
# - last_updated: Timestamp of the last update (for showing data freshness)

# From users/{userId}/course_summaries/{courseId}/holes/{holeNumber}:
# - HOLE: Hole number (for chart x-axis)
# - PAR: Par for the hole (for comparison)
# - SCORE: Score on that specific hole (key performance indicator)
# - PUTT: Number of putts on the hole (detailed performance metric)
# - YDS: Yardage of the hole (for context)
# - HDCP: Handicap of the hole (for context)
# - DD: Driving distance (if available)
# - FIR: Fairway in regulation (if available)
# - GIR: Green in regulation (if available)
# - PINDIST: Pin distance (if available)
# - 1STPU: First putt distance (if available)
# - ESC: Scrambling success (if available)

# Aggregations needed in the application logic:
# - Trend of average score over time (requires fetching multiple course_summaries or rounds)
# - Per-hole average score/putts/FIR/GIR etc. across all rounds on a specific course
# - Comparison of performance metrics across different courses

# 2. Using Firestore Real-time Listeners
# Firestore real-time listeners are ideal for near real-time updates in the user application.
# They subscribe to changes (additions, modifications, deletions) in specific documents or collections.

# To display performance charts for a specific user and course:
# - A listener can be attached to the document: users/{userId}/course_summaries/{courseId}
#   This listener will provide real-time updates to the course summary data (average score, number of rounds).
# - A listener can be attached to the collection: users/{userId}/course_summaries/{courseId}/holes
#   This listener will provide real-time updates for all hole documents for that course. This is crucial for per-hole analysis and charts.

# For broader performance overview (e.g., list of courses and their average scores for a user):
# - A listener can be attached to the collection: users/{userId}/course_summaries
#   This listener will provide updates whenever a new course summary is added or an existing one is modified for the user.

# The choice of listener scope depends on the specific chart being displayed. For a detailed view of a single course, the first two listeners are appropriate. For a user's overall performance across courses, the third listener is better.

# 3. Handling Updates from Real-time Listeners
# When a listener receives an update (a snapshot of the data):
# - The application will receive the updated data for the document or the list of documents in the collection.
# - The application's data layer will process this new data.
# - For course summary updates: The new average score, round count, etc., will update the relevant display elements.
# - For hole data updates: The application will receive changes to individual hole documents. It will need to re-aggregate the relevant metrics (like average score per hole) based on the updated set of hole documents.
# - The presentation layer (UI) will then dynamically update the charts and performance metrics displayed to the user without requiring a page refresh. This is typically handled by the UI framework (e.g., React, Vue, Angular for web; SwiftUI, Jetpack Compose for mobile) which reacts to data changes.

# 4. Optimizing Real-time Data Access
# - Limit the scope of listeners: Only listen to the documents or collections strictly necessary for the current view. For example, when viewing a specific course's details, don't listen to all course summaries for the user.
# - Use data aggregation in Firestore where possible: For metrics like average score per course, calculating and storing this in the course_summaries document reduces the need to fetch and aggregate all rounds or hole data in the application.
# - Implement pagination or infinite scrolling for large collections: If a user has played a course many times, fetching all hole data at once might be inefficient. While per-hole averages might still require all hole data, displaying a list of rounds could benefit from pagination.
# - Detach listeners when they are no longer needed (e.g., when the user navigates away from a specific course view).
# - Monitor read operations in the Firebase console to understand usage and identify potential cost or performance bottlenecks.

print("Outline for Firestore data access and real-time listeners created.")

Outline for Firestore data access and real-time listeners created.


## Implement backend reporting mechanism

### Subtask:
Design and build the process for generating HTML reports from the Firestore database. This could involve fetching data from Firestore, processing it (perhaps using pandas again), and generating HTML output. This part would likely run as a separate script or service.


**Reasoning**:
Fetch data from Firestore for report generation, process it using pandas, generate HTML reports, and save them.



In [ ]:
import pandas as pd
from datetime import datetime
import numpy as np
import firebase_admin
from firebase_admin import credentials, firestore
import os

# Assume Firebase Admin SDK is initialized and 'db' is the Firestore client.
# If not, uncomment and run the initialization code below (ensure service account key path is correct).

# service_account_key_path = "/content/drive/MyDrive/serviceAccountKey.json"
# try:
#     firebase_admin.get_app()
# except ValueError:
#     if not os.path.exists(service_account_key_path):
#         print(f"❌ Error: Service account key not found at '{service_account_key_path}'. Please check the path and ensure your Google Drive is mounted.")
#     else:
#         try:
#             cred = credentials.Certificate(service_account_key_path)
#             firebase_admin.initialize_app(cred)
#             print("✅ Firebase Admin SDK initialized successfully.")
#         except Exception as e:
#             print(f"❌ Error initializing Firebase Admin SDK: {e}")

# try:
#     db = firestore.client()
# except Exception as e:
#     print(f"❌ Error getting Firestore client: {e}")


# 1. Fetch relevant data from Firestore
print("Fetching data from Firestore for reports...")
users_ref = db.collection('users')
reports_data = []

try:
    users = users_ref.stream()
    for user_doc in users:
        user_id = user_doc.id
        course_summaries_ref = users_ref.document(user_id).collection('course_summaries')
        course_summaries = course_summaries_ref.stream()

        for course_summary_doc in course_summaries:
            course_id = course_summary_doc.id
            course_summary_data = course_summary_doc.to_dict()

            holes_ref = course_summaries_ref.document(course_id).collection('holes')
            holes = holes_ref.stream()
            hole_data_list = []
            for hole_doc in holes:
                hole_data = hole_doc.to_dict()
                hole_data['hole_number'] = hole_doc.id # Add hole number from document ID
                hole_data_list.append(hole_data)

            # Combine course summary with hole data
            report_entry = {
                'user_id': user_id,
                'course_id': course_id,
                'course_summary': course_summary_data,
                'hole_data': hole_data_list
            }
            reports_data.append(report_entry)

    print(f"✅ Fetched data for {len(reports_data)} user-course combinations.")

except Exception as e:
    print(f"❌ Error fetching data from Firestore: {e}")
    reports_data = [] # Ensure reports_data is empty on error


# 2. Process the fetched data using pandas DataFrames
if reports_data:
    print("Processing data for reports...")
    all_hole_data_df = pd.DataFrame()
    course_summaries_df = pd.DataFrame()

    processed_reports = []

    for entry in reports_data:
        user_id = entry['user_id']
        course_id = entry['course_id']
        course_summary = entry['course_summary']
        hole_data = entry['hole_data']

        # Process course summary
        summary_row = pd.DataFrame([course_summary])
        summary_row['user_id'] = user_id
        summary_row['course_id'] = course_id
        course_summaries_df = pd.concat([course_summaries_df, summary_row], ignore_index=True)

        # Process hole data
        if hole_data:
            hole_df = pd.DataFrame(hole_data)
            hole_df['user_id'] = user_id
            hole_df['course_id'] = course_id

            # Clean data: Convert relevant columns to numeric, handling errors
            numeric_cols = ['SCORE', 'PUTT', 'YDS', 'HDCP', 'PAR', 'DD', 'PINDIST', '1STPU', 'ESC']
            for col in numeric_cols:
                 if col in hole_df.columns:
                    hole_df[col] = pd.to_numeric(hole_df[col], errors='coerce')


            all_hole_data_df = pd.concat([all_hole_data_df, hole_df], ignore_index=True)

            # Example processing: Calculate average score per hole for this course
            avg_hole_scores = hole_df.groupby('hole_number')['SCORE'].mean().reset_index()
            avg_hole_scores.rename(columns={'SCORE': 'average_hole_score'}, inplace=True)


            processed_reports.append({
                'user_id': user_id,
                'course_id': course_id,
                'course_summary': course_summary, # Keep original summary for report
                'avg_hole_scores': avg_hole_scores # Add processed hole data
            })


    print("✅ Data processing complete.")
    # print("\nSample of processed hole data:")
    # display(all_hole_data_df.head())
    # print("\nSample of course summaries:")
    # display(course_summaries_df.head())

else:
    print("Skipping data processing as no data was fetched.")
    processed_reports = []


# 3. Generate HTML content for the reports
print("Generating HTML reports...")
generated_reports = {}

if processed_reports:
    for report_data in processed_reports:
        user_id = report_data['user_id']
        course_id = report_data['course_id']
        course_summary = report_data['course_summary']
        avg_hole_scores_df = report_data['avg_hole_scores']

        # Start building HTML content
        html_content = f"<h1>Performance Report for {user_id} on {course_id}</h1>"

        # Add course summary details
        html_content += "<h2>Course Summary</h2>"
        html_content += f"<p>Number of Rounds: {course_summary.get('number_of_rounds', 'N/A')}</p>"
        html_content += f"<p>Average Score: {course_summary.get('average_score', 'N/A'):.2f}</p>"
        # Add other summary fields as needed

        # Add average hole scores table
        if not avg_hole_scores_df.empty:
            html_content += "<h2>Average Score per Hole</h2>"
            # Use pandas to_html for the table
            html_content += avg_hole_scores_df.to_html(index=False)
        else:
            html_content += "<p>No hole data available for average score per hole.</p>"


        # Store the generated HTML
        report_key = f"{user_id}_{course_id}"
        generated_reports[report_key] = html_content

    print(f"✅ Generated {len(generated_reports)} HTML reports.")
else:
    print("Skipping HTML report generation as no data was processed.")


# 4. Save the generated HTML reports (example: save to files)
print("Saving HTML reports...")

output_dir = "/content/reports" # Directory to save reports
os.makedirs(output_dir, exist_ok=True)

if generated_reports:
    for report_key, html_content in generated_reports.items():
        file_path = os.path.join(output_dir, f"{report_key}_report.html")
        try:
            with open(file_path, "w") as f:
                f.write(html_content)
            print(f"  Saved report: {file_path}")
        except Exception as e:
            print(f"  ❌ Error saving report {report_key}: {e}")

    print(f"✅ HTML reports saved to '{output_dir}'.")
else:
    print("Skipping HTML report saving as no reports were generated.")


# 5. Implement error handling and logging (already included in the steps above)
print("Error handling and logging implemented in the script.")

Fetching data from Firestore for reports...
✅ Fetched data for 50 user-course combinations.
Processing data for reports...
✅ Data processing complete.
Generating HTML reports...
Skipping HTML report generation as no data was processed.
Saving HTML reports...
Skipping HTML report saving as no reports were generated.
Error handling and logging implemented in the script.


**Reasoning**:
The previous step fetched data but failed to generate reports because `processed_reports` was empty. Reviewing the code, it seems `processed_reports` is only populated if `hole_data` exists for a user-course combination. The task requires generating reports based on *all* relevant data, including the course summary even if there's no hole data. The processing logic needs to be adjusted to always include a report entry for each user-course combination and handle cases where hole data is missing.



In [ ]:
import pandas as pd
from datetime import datetime
import numpy as np
import firebase_admin
from firebase_admin import credentials, firestore
import os

# Assume Firebase Admin SDK is initialized and 'db' is the Firestore client.
# If not, uncomment and run the initialization code below (ensure service account key path is correct).

# service_account_key_path = "/content/drive/MyDrive/serviceAccountKey.json"
# try:
#     firebase_admin.get_app()
# except ValueError:
#     if not os.path.exists(service_account_key_path):
#         print(f"❌ Error: Service account key not found at '{service_account_key_path}'. Please check the path and ensure your Google Drive is mounted.")
#     else:
#         try:
#             cred = credentials.Certificate(service_account_key_path)
#             firebase_admin.initialize_app(cred)
#             print("✅ Firebase Admin SDK initialized successfully.")
#         except Exception as e:
#             print(f"❌ Error initializing Firebase Admin SDK: {e}")

# try:
#     db = firestore.client()
# except Exception as e:
#     print(f"❌ Error getting Firestore client: {e}")


# 1. Fetch relevant data from Firestore
print("Fetching data from Firestore for reports...")
users_ref = db.collection('users')
reports_data = []

try:
    users = users_ref.stream()
    for user_doc in users:
        user_id = user_doc.id
        course_summaries_ref = users_ref.document(user_id).collection('course_summaries')
        course_summaries = course_summaries_ref.stream()

        for course_summary_doc in course_summaries:
            course_id = course_summary_doc.id
            course_summary_data = course_summary_doc.to_dict()

            holes_ref = course_summaries_ref.document(course_id).collection('holes')
            holes = holes_ref.stream()
            hole_data_list = []
            for hole_doc in holes:
                hole_data = hole_doc.to_dict()
                hole_data['hole_number'] = hole_doc.id # Add hole number from document ID
                hole_data_list.append(hole_data)

            # Combine course summary with hole data
            report_entry = {
                'user_id': user_id,
                'course_id': course_id,
                'course_summary': course_summary_data,
                'hole_data': hole_data_list
            }
            reports_data.append(report_entry)

    print(f"✅ Fetched data for {len(reports_data)} user-course combinations.")

except Exception as e:
    print(f"❌ Error fetching data from Firestore: {e}")
    reports_data = [] # Ensure reports_data is empty on error


# 2. Process the fetched data using pandas DataFrames
if reports_data:
    print("Processing data for reports...")
    all_hole_data_df = pd.DataFrame()
    course_summaries_df = pd.DataFrame()

    processed_reports = []

    for entry in reports_data:
        user_id = entry['user_id']
        course_id = entry['course_id']
        course_summary = entry['course_summary']
        hole_data = entry['hole_data']

        # Process course summary
        summary_row = pd.DataFrame([course_summary])
        summary_row['user_id'] = user_id
        summary_row['course_id'] = course_id
        course_summaries_df = pd.concat([course_summaries_df, summary_row], ignore_index=True)

        # Process hole data if available
        hole_df = pd.DataFrame() # Initialize empty hole_df for this entry
        avg_hole_scores = pd.DataFrame() # Initialize empty avg_hole_scores

        if hole_data:
            hole_df = pd.DataFrame(hole_data)
            hole_df['user_id'] = user_id
            hole_df['course_id'] = course_id

            # Clean data: Convert relevant columns to numeric, handling errors
            numeric_cols = ['SCORE', 'PUTT', 'YDS', 'HDCP', 'PAR', 'DD', 'PINDIST', '1STPU', 'ESC']
            for col in numeric_cols:
                 if col in hole_df.columns:
                    hole_df[col] = pd.to_numeric(hole_df[col], errors='coerce')

            # Concatenate to the overall hole data DataFrame
            all_hole_data_df = pd.concat([all_hole_data_df, hole_df], ignore_index=True)

            # Example processing: Calculate average score per hole for this course
            if not hole_df.empty:
                # Ensure 'hole_number' is numeric for sorting if needed, but groupby works on strings too
                # Convert to numeric if it represents numbers, coerce errors to NaN
                hole_df['hole_number_numeric'] = pd.to_numeric(hole_df['hole_number'], errors='coerce')

                # Group by original 'hole_number' string, sort by the numeric version
                avg_hole_scores = hole_df.groupby('hole_number')['SCORE'].mean().reset_index()
                avg_hole_scores.rename(columns={'SCORE': 'average_hole_score'}, inplace=True)

                # Optional: Sort by hole number if they are numeric strings
                try:
                    avg_hole_scores['hole_number_numeric'] = pd.to_numeric(avg_hole_scores['hole_number'], errors='coerce')
                    avg_hole_scores = avg_hole_scores.sort_values(by='hole_number_numeric').drop(columns='hole_number_numeric')
                except ValueError:
                    pass # Keep unsorted if hole numbers are not numeric


        # Always append a processed report entry, even if hole data is empty
        processed_reports.append({
            'user_id': user_id,
            'course_id': course_id,
            'course_summary': course_summary, # Keep original summary for report
            'avg_hole_scores': avg_hole_scores # Will be empty if no hole data
        })


    print("✅ Data processing complete.")
    # print("\nSample of processed hole data:")
    # display(all_hole_data_df.head())
    # print("\nSample of course summaries:")
    # display(course_summaries_df.head())
    # print("\nSample of processed_reports structure (first entry):")
    # if processed_reports:
    #     print(processed_reports[0])


else:
    print("Skipping data processing as no data was fetched.")
    processed_reports = []


# 3. Generate HTML content for the reports
print("Generating HTML reports...")
generated_reports = {}

if processed_reports:
    for report_data in processed_reports:
        user_id = report_data['user_id']
        course_id = report_data['course_id']
        course_summary = report_data['course_summary']
        avg_hole_scores_df = report_data['avg_hole_scores']

        # Start building HTML content
        html_content = f"<h1>Performance Report for {user_id} on {course_id}</h1>"

        # Add course summary details
        html_content += "<h2>Course Summary</h2>"
        html_content += f"<p>Number of Rounds: {course_summary.get('number_of_rounds', 'N/A')}</p>"
        avg_score = course_summary.get('average_score')
        html_content += f"<p>Average Score: {avg_score:.2f}" if pd.notna(avg_score) else "<p>Average Score: N/A"
        html_content += "</p>"

        last_updated = course_summary.get('last_updated')
        if last_updated and hasattr(last_updated, 'strftime'):
             html_content += f"<p>Last Updated: {last_updated.strftime('%Y-%m-%d %H:%M:%S')}</p>"
        else:
             html_content += "<p>Last Updated: N/A</p>"


        # Add average hole scores table
        if not avg_hole_scores_df.empty:
            html_content += "<h2>Average Score per Hole</h2>"
            # Use pandas to_html for the table
            html_content += avg_hole_scores_df.to_html(index=False)
        else:
            html_content += "<p>No hole data available for average score per hole for this course.</p>"


        # Store the generated HTML
        report_key = f"{user_id}_{course_id}"
        generated_reports[report_key] = html_content

    print(f"✅ Generated {len(generated_reports)} HTML reports.")
else:
    print("Skipping HTML report generation as no data was processed.")


# 4. Save the generated HTML reports (example: save to files)
print("Saving HTML reports...")

output_dir = "/content/reports" # Directory to save reports
os.makedirs(output_dir, exist_ok=True)

if generated_reports:
    for report_key, html_content in generated_reports.items():
        file_path = os.path.join(output_dir, f"{report_key}_report.html")
        try:
            with open(file_path, "w") as f:
                f.write(html_content)
            print(f"  Saved report: {file_path}")
        except Exception as e:
            print(f"  ❌ Error saving report {report_key}: {e}")

    print(f"✅ HTML reports saved to '{output_dir}'.")
else:
    print("Skipping HTML report saving as no reports were generated.")


# 5. Implement error handling and logging (already included in the steps above)
print("Error handling and logging implemented in the script.")


Fetching data from Firestore for reports...
✅ Fetched data for 50 user-course combinations.
Processing data for reports...
✅ Data processing complete.
Generating HTML reports...
✅ Generated 50 HTML reports.
Saving HTML reports...
  Saved report: /content/reports/998b30f58115491fbf4188450158_ARRAYANES_report.html
  Saved report: /content/reports/998b30f58115491fbf4188450158_BRICEÑO_report.html
  Saved report: /content/reports/998b30f58115491fbf4188450158_BRITANIA_report.html
  Saved report: /content/reports/998b30f58115491fbf4188450158_CARMEL_report.html
  Saved report: /content/reports/998b30f58115491fbf4188450158_CORAL SPRINGS_report.html
  Saved report: /content/reports/998b30f58115491fbf4188450158_CRANDON_report.html
  Saved report: /content/reports/998b30f58115491fbf4188450158_DAVID GUTIERREZ_report.html
  Saved report: /content/reports/998b30f58115491fbf4188450158_EMERALD HILLS_report.html
  Saved report: /content/reports/998b30f58115491fbf4188450158_FUNDADORES_report.html
  Saved

## Connect data upload to real-time updates

### Subtask:
Ensure that the data upload process correctly triggers the real-time updates in the user app.


**Reasoning**:
Review the current data upload script and assess if any modifications or additional mechanisms are needed to trigger real-time updates in the user app, and then outline necessary steps. Since the previous attempts to upload data failed due to permission errors, the script needs to be re-evaluated assuming the permission issue is resolved. Firestore's real-time listeners inherently react to standard write operations, so if the app simply reflects the data as it is written, no extra steps might be needed in the upload script itself. However, if server-side logic or aggregations are required for real-time updates, Firestore triggers might be necessary.



In [ ]:
# 1. Review the current data upload script
# The current script attempts to:
# - Load data from an Excel file into pandas DataFrames (deposit_df, index_df).
# - Clean and merge the data into merged_hole_data.
# - Group the merged data by 'PLAYER' and 'CC' to represent user-course combinations.
# - Iterate through these groups.
# - For each group, it constructs Firestore paths:
#   - users/{userId}/course_summaries/{courseId} for a summary document.
#   - users/{userId}/course_summaries/{courseId}/holes/{holeNumber} for hole documents within a subcollection.
# - It calculates a course summary (number of rounds, average score) from index_df.
# - It prepares hole-by-hole data from the merged_hole_data.
# - It attempts to set the course summary document using `set(..., merge=True)`.
# - It attempts to set each hole document within the 'holes' subcollection using `set(...)`.

# 2. Consider how changes are detected by real-time listeners
# Firestore's real-time listeners attached to a document or collection will automatically
# receive updates whenever a write operation (set, update, add, delete) occurs on that
# document or within that collection.

# In this script:
# - `course_summary_ref.set(..., merge=True)` is a standard write operation on the course summary document.
#   A listener on `users/{userId}/course_summaries/{courseId}` will be triggered by this.
# - `hole_doc_ref.set(...)` is a standard write operation on each hole document.
#   A listener on the `users/{userId}/course_summaries/{courseId}/holes` collection will be triggered by these writes.

# Therefore, the current script's write operations are inherently compatible with Firestore's
# real-time listening mechanism. When this script successfully writes data after the permission
# issues are resolved, any connected user applications with active listeners on the affected
# documents or collections will automatically receive updates.

# 3. Assess if additional mechanisms (like Firestore triggers) are necessary
# Based on the description of the user app requiring "performance charts updated in real-time
# upon scorecard uploads," and assuming these charts primarily display the data being uploaded
# (course summaries and individual hole data), the standard real-time listeners should be sufficient.
# The app would listen to the relevant course_summaries and holes collections/documents and
# update its charts based on the received snapshots.

# Firestore triggers (Cloud Functions) would typically be needed if:
# - Server-side validation is required before data is considered 'final' for the app.
# - Complex aggregations or calculations need to be performed server-side upon data changes
#   and stored back in Firestore (e.g., calculating a rolling average across many rounds
#   that isn't easily done client-side).
# - Notifications or other side effects need to occur when data changes.
# - Data from multiple documents needs to be combined or transformed server-side to produce
#   the data needed for a real-time chart.

# The current script already calculates a simple average score for the course summary. If
# more complex, real-time server-side aggregations are needed (beyond what the client can
# easily compute from the streamed hole data), then Firestore triggers could be beneficial.
# However, for simply displaying the uploaded hole data and the calculated course summary,
# triggers are not strictly necessary for real-time updates. The client application can
# perform per-hole aggregations (like average score per hole across multiple rounds)
# based on the data it receives from the 'holes' subcollection listener.

# 4. Outline necessary modifications or suggest server-side components
# Based on the assessment, no *modifications to the data upload script itself* are strictly
# necessary to make the data *available* for real-time listeners, assuming the permission
# errors are resolved. The script already performs standard Firestore writes that trigger
# listeners.

# However, the outline for the overall system should *suggest* the implementation of
# real-time listeners in the user application code.

# Outline for ensuring real-time updates in the user app:

# a. In the User Application:
#    - Implement Firestore real-time listeners (`onSnapshot`) in the relevant components.
#    - For a specific course view: Attach a listener to the `users/{userId}/course_summaries/{courseId}` document
#      to get real-time updates on the summary data (average score, rounds count).
#    - For displaying hole-by-hole performance on a course: Attach a listener to the
#      `users/{userId}/course_summaries/{courseId}/holes` collection to get real-time
#      updates on all hole documents for that course.
#    - Implement logic in the app's data layer to process the snapshots received from listeners.
#      This includes updating state with new/modified/deleted documents.
#    - Implement logic to perform client-side aggregations (e.g., calculating average
#      score per hole across all rounds currently available in the streamed data) based
#      on the data from the 'holes' collection listener.
#    - Ensure the UI components (charts, metrics display) react to changes in the app's
#      data state, automatically re-rendering with the new data.
#    - Properly detach listeners when views are no longer active to prevent memory leaks
#      and unnecessary billing.

# b. Consider Firestore Triggers (Optional, based on future complexity):
#    - If complex, server-side data processing or aggregation is needed upon each hole
#      or round upload (e.g., calculating user handicaps that depend on multiple rounds),
#      implement Cloud Functions triggered by Firestore writes (`onCreate`, `onUpdate`)
#      on the `users/{userId}/course_summaries/{courseId}` or
#      `users/{userId}/course_summaries/{courseId}/holes` paths.
#    - These triggers would perform the necessary server-side logic and potentially write
#      the results back to Firestore (e.g., updating the user's main profile document
#      with their current handicap), which would, in turn, trigger listeners on those
#      documents in the user app.

# Conclusion: The existing upload script's write operations are suitable for triggering
# real-time listeners. The focus for ensuring real-time updates should be on implementing
# the listener logic correctly within the user application. Server-side triggers are an
# option for more complex real-time processing but not strictly required for basic data
# reflection and client-side aggregation of uploaded data.

print("Assessment of real-time update triggering complete.")
print("Outline for user app real-time data handling created.")

Assessment of real-time update triggering complete.
Outline for user app real-time data handling created.


## Integrate backend reporting into workflow

### Subtask:
Determine how and when the backend reports will be generated (e.g., scheduled task, manual trigger) and where they will be stored or accessed.


**Reasoning**:
Determine how and when the backend reports will be generated, where they will be stored, and how they will be accessed, then finish the subtask.



In [ ]:
# Determine the report generation trigger and frequency
# Option 1: Manual Trigger (Simplest for "practice mode")
# - Reports are generated only when the backend script is manually executed (e.g., in Colab).
# - Suitable for testing and infrequent report needs.

# Option 2: Scheduled Task
# - Reports are generated automatically on a predefined schedule (e.g., daily, weekly).
# - Requires a scheduling mechanism.
# - Suitable for regular reporting needs.

# Option 3: Event Trigger (e.g., upon data upload)
# - Reports are generated automatically whenever new data is uploaded to Firestore.
# - Requires a serverless function triggered by Firestore writes.
# - Suitable for near real-time reporting, but can be costly with frequent uploads.

# For a "practice mode" and initial setup, a Manual Trigger or a simple Scheduled Task is most appropriate.
# Let's plan for a Scheduled Task as it provides a more automated workflow for ongoing use,
# while still being manageable for development.

# Outline the infrastructure for report generation
# If using a Scheduled Task:
# - Google Cloud Scheduler: To define the schedule (e.g., run daily at a specific time).
# - Cloud Run or Cloud Functions: To execute the report generation script.
#   - Cloud Run is suitable for longer-running tasks or containerized applications.
#   - Cloud Functions are good for shorter, event-driven tasks.
#   - Given the potential time to fetch and process data for many users/courses, Cloud Run might be more robust.
# - The Python script developed in the previous step would be deployed to Cloud Run.

# Determine where the generated HTML reports will be stored
# Options:
# - Google Cloud Storage: Cost-effective, scalable storage for files. Reports can be stored as HTML files.
# - Serve directly from a web server: Requires a server to host the files.
# - Store back in Firestore: Not recommended for large HTML content due to document size limits and cost.

# Google Cloud Storage is the recommended approach for storing generated HTML files.

# Describe how users or other systems will access the generated reports
# If stored in Google Cloud Storage:
# - Generate signed URLs: For temporary, secure access to individual report files.
# - Set up public access (with caution): If reports are not sensitive and can be publicly available (less likely for personal performance data).
# - Integrate with a web application: A backend API in the user-facing application can fetch signed URLs from Cloud Storage and redirect the user or serve the content.

# Summarize the chosen workflow for report generation, storage, and access

# Chosen Workflow:
# 1. Trigger: Scheduled Task using Google Cloud Scheduler (e.g., daily).
# 2. Execution Environment: Python script deployed to Google Cloud Run.
# 3. Data Source: Fetch data from Firestore using the Firebase Admin SDK.
# 4. Processing and Generation: Use pandas within the script to process data and generate HTML content.
# 5. Storage: Store generated HTML files in a dedicated Google Cloud Storage bucket.
# 6. Access: User-facing application backend generates signed URLs for users to access their specific reports from Cloud Storage.

print("Report generation, storage, and access plan outlined.")

Report generation, storage, and access plan outlined.


## Summary:

### Data Analysis Key Findings

*   The existing Firestore data structure (`/users/{userId}/course_summaries/{courseId}` and `/users/{userId}/course_summaries/{courseId}/holes/{holeNumber}`) is deemed suitable for both real-time user performance display and backend report generation, although manual verification via the Firebase console is recommended.
*   Firestore's native real-time listeners are the intended mechanism for providing real-time performance chart updates in the user application upon data uploads, as the standard write operations used in the upload script inherently trigger these listeners.
*   A backend process has been successfully designed and implemented to fetch performance data from Firestore, process it using pandas (including calculating average scores per hole), generate HTML reports for each user-course combination, and save these reports as individual files.
*   The backend report generation workflow is planned to be triggered by a scheduled task using Google Cloud Scheduler, executed on Google Cloud Run, with reports stored in Google Cloud Storage and accessed by users via signed URLs generated by the application backend.

### Insights or Next Steps

*   Proceed with the implementation of Firestore real-time listeners (`onSnapshot`) within the user-facing application to display dynamic performance charts based on the data structure and upload process.
*   Set up and deploy the backend report generation script to Google Cloud Run, configure a scheduled trigger using Google Cloud Scheduler, and establish the Google Cloud Storage bucket for report storage, including implementing the signed URL generation logic in the application backend for secure access.


In [ ]:
service_account_key_path = '/content/drive/MyDrive/serviceAccountKey.json'

try:
    with open(service_account_key_path, 'r') as f:
        key_file_content = f.read()
    print(key_file_content)
except FileNotFoundError:
    print(f"Error: Service account key not found at '{service_account_key_path}'. Please check the path.")
except Exception as e:
    print(f"Error reading service account key file: {e}")

{
  "type": "service_account",
  "project_id": "dnashot-b2fa0",
  "private_key_id": "a01c4f244b8966f3d570fbffe8e57ffe9083807a",
  "private_key": "-----BEGIN PRIVATE KEY-----\nMIIEvAIBADANBgkqhkiG9w0BAQEFAASCBKYwggSiAgEAAoIBAQC97j626vSXluSg\ntHcDDSJUJoYOxij0TTdjGBdH2Rc9SLzQQkjBTWskagcAZCFLBamUKalIJfYKNv9g\nuZsvap2IyEAUUhp0A9XWmd9ynfkvCR5mUxrSdJbmDvuvPjAkAZqmCNC1XJxgvAfS\njmDmu993FSYcQF37wFaZ5aYys0CVMBDZRx0F51z8e5B7iKTS5lDBiHnF340noRj+\nPDWe+SvlSotE3Dn5RQEAlpl5XkbY6CHvTKCIb7T/hNJqM2nSaglIMwCZb2O3fDp3\nUjQ2zqVUU8eNyGdMGPDCFFVMwH1yIFrn+a3VECFASAsnXNAEG4Cxg1DQ6GbIpL+H\n5AirugPRAgMBAAECgf8DSwsyMJnKXFknMIMfnBr54bedsDM19OtyZmjbQ4Owo/J8\nrY6Vk5ywQml6NF0D1nd4mpH1JFmABoF6jmrnPWkDrZLRtBHX6OXynJrCRTmlH1sb\nxnk9xnGDC8IjJEI/GI7b9cjHZ1TSI2kHJorVxmaadqRX5d4X1kPt0RnfYacmdElF\nqEUWJkKz6mjASZy98IJrlRk66rXRh30KqjMWXRvfAnhwq829+AurMAT7BXvme8Ai\nNGOHrPpr2ZequogRsHaK53J+AhGmPZbv7NUx3yo/mqloBmvJd8lPRp9bBVjyM020\n2f+TJ4wiK6HNFgtUfN4S6t18ykCAf70qGFV836ECgYEA6K1RAHKicrl7fF2QRY7m\n4ZObW7t/4MxwnQGPt9TNa08RQ2T8wuaXr